### DISCUSSION TITLE
How to capture/bind to "shift+enter"

### DISCUSSION THREAD

**Discussion by jspv** [Textualize/textual#3542] [Q&A]
I'm having difficulty either binding or capturing via on_key the shift+enter combination.  It appears to always come in to textual as an "enter".  Is there a way to do this?  

This can be seen via the `textual keys` tool.

**Comment by davep** [comment]
When you say this can be seen via `textual keys`, I take it you mean that no matter the combination pressed, you only see <kbd>Enter</kbd>?

I've recently been testing a lot of terminals and seeing what keys are recognised; I don't believe I've seen one that processed <kbd>Shift</kbd>+<kbd>Enter</kbd> as anything other than <kbd>Enter</kbd>. As it stands right now, if the terminal emulator you're using doesn't tell Textual that a different key combination was pressed, Textual can't know.

#3411 might be interesting here but it's not something we've looked into yet, and I don't personally know if it would actually allow this.

**Comment by jspv** [comment]
yes, in `textual keys` only 'enter' is displayed when shift+enter is used.  Your explanation makes sense except though when I'm using vim in the same terminal (iterm in my case), shift+enter definitely exhibits different behavior than enter; so I believe it is sending a different set of codes.   There is some discussion on this here:  https://stackoverflow.com/questions/16359878/how-to-map-shift-enter

**Reply by davep** [reply]
The discussion there seems to suggest that it requires specific configuration in different terminals to make it work. Am I misreading?

**Reply by jspv** [reply]
No, I read it similarly, just noting that iTerm2 seems to map it natively and some other terminals very well may as well as that post was 10 years ago.

**Comment by ZynthCode** [comment]
I had the same issues, so I made a quick hack that works surprisingly well for only **Windows 11 in PowerShell** (don't use this unless you are only using this for PowerShell):

```py
from textual.widgets import TextArea
import time

class InputTextField(TextArea):

    default_text_height = 5
    DEFAULT_CSS = """
    InputTextField {
        height: 5;
    }
    """

    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)
        self.last_key_time = None
        self.last_key = None

    def _on_key(self, event) -> None:
        current_time = time.time()
        key = event.key

        # Shift+Enter (Tested on Windows terminal, have not verified linux or mac)
        if self.last_key == "f12" and key == "c" and self.last_key_time and current_time - self.last_key_time < 0.01: 
            self.insert("\n")
            event.prevent_default()

        elif (key == "enter"):
            event.prevent_default()
            # TODO: Submit content here
            self.styles.height = self.default_text_height

        self.last_key = key
        self.last_key_time = current_time

```

**Reply by davep** [reply]
As mentioned earlier in this discussion: in many terminals in many environments, pressing shift+enter will just result in enter being sent to the application, if you've got a way of making it work in the Windows Terminal it really will be a hack, and possibly even one that might stop working in another release of Windows Terminal.

**Reply by majkelx** [reply]
In the same terminal in the same emulator (1. MacOS Terminal, 2. Terminus), other tools (e.g. Claude) recognize and distinguish SHIFT-ENTER and ENTER, so you can not blame "terminals" and "environments".

**Reply by willmcgugan** [reply]
@majkelx You are commenting on a 2 year old closed discussion.
